### PR TITLE
Jwt auth

### PR DESCRIPTION
**Pull request by piercetrey-figure**
* Swaps out custom in-request signature for jwt header-based auth scheme
* Can accept an existing jwt generated via Provenance Wallet or the like, or generate one either per-request or on a reusable basis when provided a KeyPair
* Server-side uses a Grpc interceptor to inspect/validate the jwt on an incoming request and reject the request if unset/invalid, or set the requestor's public key/address in context for retrieval within the service

### CODE REVIEW

File: client/src/main/kotlin/io/provenance/objectstore/gateway/client/GatewayClient.kt [hunk 1]
```diff
@@ -31,28 +41,42 @@ class GatewayClient(val config: ClientConfig): Closeable {
...
+        .withClaim("addr", keyPair.public.getAddress(config.mainNet))
+        .sign(Algorithm.ECDSA256K(keyPair.public as ECPublicKey, keyPair.private as ECPrivateKey))
+
+    /**
+     * Fetch scope data from gateway, creating a fresh JWT for authentication
+     * @param scopeAddress the scope's address
+     * @param keyRef the KeyRef of the key to sign the request with/decrypt the received response
+     * @param timeout an optional timeout for the request/used in the request to expire signature
+     */
+    fun requestScopeData(scopeAddress: String, keyPair: KeyPair, timeout: Duration = Duration.ofSeconds(10)): GatewayOuterClass.FetchObjectResponse {
```
Comment (hyperschwartz) [inline client/src/main/kotlin/io/provenance/objectstore/gateway/client/GatewayClient.kt:58]: This is nice and easy to use! Excellent

File: server/src/main/kotlin/io/provenance/objectstore/gateway/service/JwtVerificationService.kt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        val verified = try {
+            JWT.require(algorithm)
+                .acceptExpiresAt(0L)
+                .build()
+                .verify(jwt)
+        } catch (e: JWTVerificationException) {
+            throw JwtValidationException(e.message)
+        }
+
+        if (!verified.claims.containsKey(Constants.JWT_EXPIRATION_KEY)) {
```
Comment (hyperschwartz) [inline server/src/main/kotlin/io/provenance/objectstore/gateway/service/JwtVerificationService.kt:39]: Does this blow up above if the expiration time has already passed?  If not, might need to add a check in here that OffsetDateTime.now() <= expiration time.  Likely will include timezone screwery, though lol

Comment (hyperschwartz) [reply server/src/main/kotlin/io/provenance/objectstore/gateway/service/JwtVerificationService.kt:39]: Ah, I see the test that proves that this is a thing.  Don't mind me :disappear:

**Review by hyperschwartz — APPROVED** [review]

File: server/src/main/kotlin/io/provenance/objectstore/gateway/service/JwtVerificationService.kt (hunk 2 above)
Comment (piercetrey-figure) [reply server/src/main/kotlin/io/provenance/objectstore/gateway/service/JwtVerificationService.kt:39]: yeah, apparently the expiration is optional and if missing the verification doesn't care, and there didn't seem to be an option to force it to care. Seemed important that we don't allow non-expiring jwts though

File: server/src/main/kotlin/io/provenance/objectstore/gateway/service/JwtVerificationService.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        val algorithm = Algorithm.ECDSA256K(publicKey as ECPublicKey, null)
+
+        val verified = try {
+            JWT.require(algorithm)
+                .acceptExpiresAt(0L)
+                .build()
+                .verify(jwt)
+        } catch (e: JWTVerificationException) {
+            throw JwtValidationException(e.message)
```
[the remark this answers is not in the document]
Comment (piercetrey-figure) [reply server/src/main/kotlin/io/provenance/objectstore/gateway/service/JwtVerificationService.kt]: good idea, I just added it to the Status's cause... I don't believe that gets transmitted to the client, but could be helpful in debugging things server side for sure

File: client/src/main/kotlin/io/provenance/objectstore/gateway/client/GatewayClient.kt (hunk 1 above)
Comment (piercetrey-figure) [reply client/src/main/kotlin/io/provenance/objectstore/gateway/client/GatewayClient.kt:58]: hopefully you will only have to change from a `KeyRef` to a `KeyPair` from the verifier... unfortunately there didn't seem to be any way that I could use our signer to sign the JWT, so I needed access to the private key

Comment (piercetrey-figure) [reply client/src/main/kotlin/io/provenance/objectstore/gateway/client/GatewayClient.kt:66]: nevermind on that, figured out a way to add in my own Algorithm class to use for generating the JWT, so we can deal with a KeyRef still, so the function signature can stay the same

**Pull request merged**